File: lc/go/happy-number.go
// 1 = 1^2 = 1

// 4 5 3 4 ... stop -- not happy
// 8 9 ... 1 ... happy
// 1 
// need a hashset to know if we saw the number before...
// 2 = 4
// 4 -- 16
// 1 6 -- 37
// 3 7 -- 9 + 49 = 58
// 5 8 -- 25 + 64 = 89
// 8 9 -- 72 + 81 = 159

type nothing struct{}
func isHappy(n int) bool {
    seen := map[int]nothing {}
    for ; n != 1; {
        sum := 0
        for ; n != 0; {
            sum += (n % 10) * (n % 10)
            n /= 10
        }
        if _, ok := seen[sum]; ok { return false }
        seen[sum] = nothing{}
        n = sum

    }
    return true
}

File: lc/py/happy-number.py
# need a set: if you see something youve seen b4, return false
# if you hit 1, return true
# 19
# 1 + 9 = 82
# 8 + 2 = 68

# mod, n |  1 1
# ss 1
# have (100 1)
# agg the next n as per the square, until n == 0. then make n the ss -- sum of squares
class Solution:
    def isHappy(self, n: int) -> bool:
        have = set()
        while n != 1:
            ss = 0
            while n != 0:
                mod, n = n % 10, n // 10
                ss += mod * mod
            n = ss
            if n in have:
                return False
            have.add(n)
        return True